drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,746 @@
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+
+	i2c_base_addr->CON = 0;
+	i2c_base_addr->PSC = data->speed_config.pscstate;
+	i2c_base_addr->SCLL = data->speed_config.scllstate;
+	i2c_base_addr->SCLH = data->speed_config.sclhstate;
+	i2c_base_addr->CON = I2C_OMAP_CON_EN;
+	uint16_t iestate = (I2C_OMAP_IE_XRDY | I2C_OMAP_IE_RRDY | I2C_OMAP_IE_ARDY |
+			    I2C_OMAP_IE_NACK | I2C_OMAP_IE_AL) |
+			   ((data->current_msg.len) ? (I2C_OMAP_IE_RDR | I2C_OMAP_IE_XDR) : 0);
+	i2c_base_addr->IRQENABLE_SET = iestate;
+}
+
+/**
+ * @brief Reset the OMAP I2C controller.
+ *
+ * This function resets the OMAP I2C controller specified by the device pointer.
+ *
+ * @param dev Pointer to the device structure for the I2C controller.
+ * @return 0 on success, negative errno code on failure.
+ */
+static int i2c_omap_reset(const struct device *dev)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint64_t timeout;
+	uint16_t sysc;
+
+	sysc = i2c_base_addr->SYSC;
+	i2c_base_addr->CON &= ~I2C_OMAP_CON_EN;
+	timeout = k_uptime_get() + I2C_OMAP_TIMEOUT;
+	i2c_base_addr->CON = I2C_OMAP_CON_EN;
+	while (!(i2c_base_addr->SYSS & SYSS_RESETDONE_MASK)) {
+		if (k_uptime_get() > timeout) {
+			LOG_WRN("timeout waiting for controller reset");
+			return -ETIMEDOUT;
+		}
+		k_busy_wait(100);
+	}
+	i2c_base_addr->SYSC = sysc;
+	data->bb_valid = 0;
+	return 0;
+}
+
+/**
+ * @brief Set the speed of the OMAP I2C controller.
+ *
+ * This function sets the speed of the OMAP I2C controller based on the
+ * specified speed parameter. The speed can be set to either Fast mode or
+ * Standard mode.
+ *
+ * @param dev The pointer to the device structure.
+ * @param speed The desired speed for the I2C controller.
+ *
+ * @return 0 on success, negative error code on failure.
+ */
+static int i2c_omap_set_speed(const struct device *dev, uint32_t speed)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+
+	/* If configured for High Speed */
+	switch (speed) {
+	case I2C_BITRATE_FAST:
+		/* Fast mode */
+		data->speed_config = (struct i2c_omap_speed_config){
+			.pscstate = 9,
+			.scllstate = 7,
+			.sclhstate = 5,
+		};
+		break;
+	case I2C_BITRATE_STANDARD:
+		/* Standard mode */
+		data->speed_config = (struct i2c_omap_speed_config){
+			.pscstate = 23,
+			.scllstate = 13,
+			.sclhstate = 15,
+		};
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Configure the OMAP I2C controller with the specified device configuration.
+ *
+ * This function configures the OMAP I2C controller with the specified device configuration.
+ *
+ * @param dev The pointer to the device structure.
+ * @param dev_config The device configuration to be applied.
+ *
+ * @return 0 on success, negative error code on failure.
+ */
+static int i2c_omap_configure(const struct device *dev, uint32_t dev_config)
+{
+	uint32_t speed_cfg;
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+
```

Comment:
There should be a semaphore used here as a lock to order access by threads

Reply:
I have updated it.